[clang-tidy] Add modernize-make-direct check

hjanuschka opened the pull request:
Fixes #121295.

Adds a new check that converts std::make_* function calls to direct constructor calls using CTAD. Transforms make_optional, make_unique, make_shared and make_pair into their equivalent direct constructor calls, leveraging C++17's class template argument deduction.

hjanuschka committed:
[clang-tidy] Add modernize-make-direct check

Adds a new check that converts std::make_* function calls to direct constructor
calls using CTAD. Transforms make_optional, make_unique, make_shared and make_pair
into their equivalent direct constructor calls, leveraging C++17's class template
argument deduction.

firewave commented:
This will conflict with `modernize-make-shared` and `modernize-make-unique`.

I also very sure having `new` any modern C++ code is very much to be avoided.

Having no insight on the differences of the inner workings - but just based on https://en.cppreference.com/w/cpp/memory/shared_ptr/make_shared#Notes this appears to be something behaving very differently and less optimal/safe.

hjanuschka committed:
Address PR feedback and refactor to TransformerClangTidyCheck

- Refactor implementation to use TransformerClangTidyCheck for cleaner code
- Remove make_unique and make_shared transformations to avoid conflicts and new usage
- Add configuration options: CheckMakePair, CheckMakeOptional, CheckMakeTuple
- Add support for std::make_tuple transformations
- Update documentation with options and warnings about smart pointer functions
- Fix clang-format issues in ModernizeTidyModule.cpp
- Add C++17 requirement enforcement with proper test setup
- All tests passing with correct CTAD transformations

hjanuschka committed:
Fix CI failures: clang-format and RST title underline

- Apply clang-format fixes to MakeFunctionToDirectCheck.cpp
- Fix RST title underline length in modernize-make-direct.rst documentation
- All tests passing after fixes

hjanuschka committed:
Fix Sphinx documentation: rename file and fix reference

- Rename modernize-make-direct.rst to make-direct.rst to follow convention
- Fix ReleaseNotes.rst document reference path
- This resolves 'unknown document' Sphinx errors in CI

vbvictor reviewed (changes requested):
Generally good check idea, but needs some polishing.
- Neither `modernize-make-function-to-direct` nor `modernize-make-direct` sounds clear to me (specifically `direct` part), the main purpose of this transformation is to use `CTAD`, it should be in the check name at least. `modernize-use-ctad-constructor` could be it, but I'm not 100% sure others will agree.
- I don't like having many options that all do the same job. How about a list of key-value pairs of classes and their corresponding `make_` function. Something like "std::pair=std::make_pair;std::optional=std::make_optional...". Look at `readability-suspicious-call-argument.Abbreviations` for reference.
- Enhance tests with type aliases, macros, template parameters

carlosgalvezp commented on clang-tools-extra/docs/clang-tidy/checks/modernize/make-direct.rst:
```diff
@@ -0,0 +1,46 @@
+Options
+-------
+
+.. option:: CheckMakePair
```
This should instead be 1 option with a semicolon separated list of which functions to transform.

carlosgalvezp commented on clang-tools-extra/docs/clang-tidy/checks/modernize/make-direct.rst:
```diff
@@ -0,0 +1,46 @@
+      exception safety that would be lost with direct construction.
+   3. Direct use of ``new`` is discouraged in modern C++ code.
+   
+   Use the dedicated ``modernize-make-unique`` and ``modernize-make-shared`` checks
```
This is confusing, since these checks do the exact opposite of what this check is doing. I would vote for removing this line.

carlosgalvezp reviewed (changes requested):
Couple of minor changes.

I'm not convinced about the name of the check, `make-function-to-direct` does not really say much. What is a "make function" and what is "direct"?

Would it make sense to call it `modernize-use-ctad` and possibly leave the door open for more use cases in follow-up patches?

PS: please use the `rename_check.py` script to rename the check, to ensure all places get updated. But let's wait until a consensus is reached before churning the code with name changes.

vbvictor commented:
> Would it make sense to call it `modernize-use-ctad` and possibly leave the door open for more use cases in follow-up patches?

+1 on `modernize-use-ctad`, good to leave space for more patches.

vbvictor commented:
Did it actually land as https://github.com/llvm/llvm-project/commit/c741fffc1aa978e39946bb34efc455dc9333f993..?
It doesn't have PR linked and original issue is still open.
Also, it's not listed on https://clang.llvm.org/extra/clang-tidy/checks/list.html
